File: ruby/lib/enolib/elements/element_base.rb
# frozen_string_literal: true

module Enolib
  class ElementBase
    def initialize(context, instruction, parent = nil)
      @context = context
      @instruction = instruction
      @parent = parent
    end

    def comment_error(message = nil)
      if block_given?
        message = yield(@context.comment(@instruction))
      elsif message.is_a?(Proc)
        message = message.call(@context.comment(@instruction))
      end

      unless message
        raise ArgumentError, 'A message or message function must be provided'
      end

      Errors::Validation.comment_error(@context, message, @instruction)
    end

    def error(message = nil)
      if block_given?
        message = yield(self)  # Revisit self in this context - problematic
      elsif message.is_a?(Proc)
        message = message.call(self)  # Revisit self in this context - problematic
      end

      unless message
        raise ArgumentError, 'A message or message function must be provided'
      end

      Errors::Validation.element_error(@context, message, @instruction)
    end

    def key(loader = nil)
      loader = Proc.new if block_given?

      @touched = true

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      begin
        loader.call(_key)
      rescue => message
        raise Errors::Validation.key_error(@context, message, @instruction)
      end
    end

    # TODO: All implementations - this could be called on the document (?),
    # via both element and section, most likely triggering erratic behavior
    # because the error->selection chain does not handle the missing key range?
    # (Also being able to call it and doing so does not make sense in the first place)
    def key_error(message = nil)
      if block_given?
        message = yield(_key)
      elsif message.is_a?(Proc)
        message = message.call(_key)
      end

      unless message
        raise ArgumentError, 'A message or message function must be provided'
      end

      Errors::Validation.key_error(@context, message, @instruction)
    end

    def optional_comment(loader = nil)
      loader = Proc.new if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _comment(loader, required: false)
    end

    def optional_string_comment
      _comment(required: false)
    end

    def raw
      @context.raw(@instruction)
    end

    def required_comment(loader = nil)
      loader = Proc.new if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _comment(loader, required: true)
    end

    def required_string_comment
      _comment(required: true)
    end

    def string_key
      @touched = true

      _key
    end

    def touch
      @touched = true
    end

    private

    def _comment(loader = nil, required:)
      @touched = true

      comment = @context.comment(@instruction)

      if comment
        return comment unless loader

        begin
          loader.call(comment)
        rescue => message
          raise Errors::Validation.comment_error(@context, message, @instruction)
        end
      else
        return nil unless required

        raise Errors::Validation.missing_comment(@context, @instruction)
      end
    end

    def _key
      return nil if @instruction[:type] == :document
      return @instruction[:parent][:key] if @instruction[:type] == :list_item

      @instruction[:key]
    end
  end
end

File: python/enolib/elements/element_base.py
from ..errors.validation import Validation
from ..constants import DOCUMENT, LIST_ITEM

class ElementBase:
    def __init__(self, context, instruction, parent=None):
        self._context = context
        self._instruction = instruction
        self._parent = parent  # TODO: Could be set optionally as well (needs hasattr checks everywhere then)

    def _comment(self, loader=None, *, required):
        self._touched = True

        comment = self._context.comment(self._instruction)

        if not comment:
            if required:
                raise Validation.missing_comment(self._context, self._instruction)
            else:
                return None

        if not loader:
            return comment

        try:
            return loader(comment)
        except ValueError as message:
            raise Validation.comment_error(self._context, message, self._instruction)

    def _key(self):
        if self._instruction['type'] is DOCUMENT:
            return None

        if self._instruction['type'] is LIST_ITEM:
            return self._instruction['parent']['key']

        return self._instruction['key']

    def comment_error(self, message):
        if callable(message):
            message = message(self._context.comment(self._instruction))

        return Validation.comment_error(self._context, message, self._instruction)

    def error(self, message):
        if callable(message):
            message = message(self)  # Revisit self in this context - problematic

        return Validation.element_error(self._context, message, self._instruction)

    def key(self, loader):
        self._touched = True

        try:
            return loader(self._key())
        except ValueError as message:
            raise Validation.key_error(self._context, message, self._instruction)

    def key_error(self, message):
        if callable(message):
            message = message(self._key())

        return Validation.key_error(self._context, message, self._instruction)

    def optional_comment(self, loader):
        return self._comment(loader, required=False)

    def optional_string_comment(self):
        return self._comment(required=False)

    def raw(self):
        return self._context.raw(self._instruction)

    def required_comment(self, loader):
        return self._comment(loader, required=True)

    def required_string_comment(self):
        return self._comment(required=True)

    def string_key(self):
        self._touched = True

        return self._key()

    def touch(self):
        self._touched = True
